kore/test/Test/Kore/AllPath.hs

```diff
@@ -0,0 +1,347 @@
+                 ( HasCallStack )
+
+import qualified Kore.AllPath as AllPath
+import qualified Kore.Step.Representation.MultiOr as MultiOr
+import qualified Kore.Step.Strategy as Strategy
+import           Kore.Step.Transition
+                 ( runTransitionT )
+import qualified Kore.Step.Transition as Transition
+
+import Test.Kore.Comparators ()
+import Test.Tasty.HUnit.Extensions
+import Test.Terse
+
+type ExecutionGraph = Strategy.ExecutionGraph (AllPath.ProofState Integer) ()
+
+emptyExecutionGraph :: AllPath.ProofState Integer -> ExecutionGraph
+emptyExecutionGraph = Strategy.emptyExecutionGraph
+
+insNode
+    :: (Gr.Node, AllPath.ProofState Integer)
+    -> ExecutionGraph
+    -> ExecutionGraph
+insNode = Strategy.insNode
+
+insEdge
+    :: (Gr.Node, Gr.Node)
+    -> ExecutionGraph
+    -> ExecutionGraph
+insEdge = Strategy.insEdge
+
+test_unprovenNodes :: [TestTree]
+test_unprovenNodes =
+    [ AllPath.unprovenNodes
+        (emptyExecutionGraph AllPath.Proven)
+        `satisfies_`
+        Foldable.null
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 1))
+        `satisfies_`
+        (not . Foldable.null)
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 1))
+        `equals`
+        (MultiOr.MultiOr [1])
+        $  "returns single unproven node"
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insNode (2, AllPath.Proven)
+        )
+        `equals_`
+        (MultiOr.MultiOr [0, 1])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Proven)
+            & insEdge (0, 2)
+        )
+        `equals_`
+        (MultiOr.MultiOr [1])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Goal 2)
+            & insEdge (1, 2)
+            & insNode (3, AllPath.Proven)
+            & insEdge (2, 3)
+        )
+        `equals_`
+        (MultiOr.MultiOr [])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.GoalRem 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Proven)
+            & insEdge (0, 2)
+        )
+        `equals_`
+        (MultiOr.MultiOr [1])
+    ]
+
+-- | Simple program configurations for unit testing.
+data K = BorC | A | B | C | D | E | F | Bot
+    deriving (Eq, Ord, Show)
+
+matches :: K -> K -> Bool
+matches B BorC = True
+matches C BorC = True
+matches a b    = a == b
+
+instance EqualWithExplanation K where
+    compareWithExplanation = rawCompareWithExplanation
+    printWithExplanation = show
+
+type Goal = (K, K)
+
+type ProofState = AllPath.ProofState Goal
+
+type Rule = (K, K)
+
+ruleAC :: Rule
+ruleAC = (A, C)
+
```

Comment:
I'm not sure the name is preferable to the definition - especially since you later use plain tuples like  `unmodified (AllPath.Goal    (A, B))`.

Reply:
I think it obscures whether the tuple is a `Rule` or a `Goal`, but I can use a `newtype` for that instead.